tools/statplot.py

```diff
@@ -183,51 +189,62 @@ def KeyPressed(self, widget, event):
+                    self.get_scale = self.ax.get_xscale
+                    curData = self.xData
+                elif key == 'y':
+                    self.get_scale = self.ax.get_yscale
+                    curData = self.yData
+                if (self.get_scale() == 'linear'
+                        and (curData == 0).all()):
                     warnings.warn('Change to logarithmic scale denied: the '
-                                  + 'selected variable for the ' + key + ' '
-                                  + 'axis is a constant.', stacklevel=2)
+                                  + 'selected variable for the ' + key
+                                  + ' axis is null.', stacklevel=2)
                     scale = 'linear'
-            elif self.get_scale() in ['log', 'symlog']:
-                scale = 'linear'
-            self.FormatAxis(key, scale)
-            self.fig.canvas.draw()
+                elif (self.get_scale() == 'linear'
+                      and max(abs(curData)) / min(abs(curData)) < 10):
+                    warnings.warn('Change to logarithmic scale denied: the '
+                                  + 'selected variable for the ' + key
+                                  + ' axis has a range of variation smaller '
+                                  + 'than one order of magnitude.',
+                                  stacklevel=2)
+                    scale = 'linear'
+                elif self.get_scale() == 'linear':
+                    axMin = min(abs(curData[curData != 0]))
+                    axMax = max(abs(curData))
+                    if axMin != axMax and min(curData) < 0:
+                        scale = 'symlog'
+                    elif axMin != axMax:
+                        scale = 'log'
+                    else:
+                        warnings.warn('Change to logarithmic scale denied: '
+                                      + 'the selected variable for the ' + key
+                                      + ' axis is a constant.', stacklevel=2)
+                        scale = 'linear'
+                elif self.get_scale() in ['log', 'symlog']:
+                    scale = 'linear'
+                self.FormatAxis(key, scale)
+                self.fig.canvas.draw_idle()
+                self.fig.canvas.draw()
+                self.fig.canvas.flush_events()
         elif key == 'l':
-            self.PlotType = 'line' if self.PlotType == 'marker' else 'marker'
-            self.PlotData('update', self.PlotType, self.ax.get_xscale(),
-                          self.ax.get_yscale())
+            if self.values.ndim == 1:
+                warnings.warn('Insufficient data available to turn on line '
+                              + 'display', stacklevel=2)
+            else:
+                self.PlotType = 'marker' if self.PlotType == 'line' else 'line'
+                self.PlotData('update', self.PlotType, self.ax.get_xscale(),
+                              self.ax.get_yscale())
```

Comment:
Early return pattern here too?

```suggestion
            if self.values.ndim == 1:
                warnings.warn('Insufficient data available to turn on line '
                              + 'display', stacklevel=2)
                return

            self.PlotType = 'marker' if self.PlotType == 'line' else 'line'
            self.PlotData('update', self.PlotType, self.ax.get_xscale(),
                          self.ax.get_yscale())
```